# Workbook

Use this notebook to complete the exercises throughout the workshop.

#### Table of Contents
- [Section 1 &ndash; Getting Started with Pandas](#Section-1)
- [Section 2 &ndash; Data Wrangling](#Section-2)
- [Section 3 &ndash; Data Visualization](#Section-3)

--- 

### Section 1

#### Exercise 1.1
##### Create a DataFrame by reading in the `2019_Yellow_Taxi_Trip_Data.csv` file. Examine the first 5 rows.

In [26]:
import pandas as pd

yellow_taxi_trip = pd.read_csv('../data/2019_Yellow_Taxi_Trip_Data.csv')

df = pd.DataFrame(yellow_taxi_trip)
df.head(5)


vendorid     tpep_pickup_datetime    tpep_dropoff_datetime   
0         2  2019-10-23T16:39:42.000  2019-10-23T17:14:10.000  \
1         1  2019-10-23T16:32:08.000  2019-10-23T16:45:26.000   
2         2  2019-10-23T16:08:44.000  2019-10-23T16:21:11.000   
3         2  2019-10-23T16:22:44.000  2019-10-23T16:43:26.000   
4         2  2019-10-23T16:45:11.000  2019-10-23T16:58:49.000   

   passenger_count  trip_distance  ratecodeid store_and_fwd_flag   
0                1           7.93           1                  N  \
1                1           2.00           1                  N   
2                1           1.36           1                  N   
3                1           1.00           1                  N   
4                1           1.96           1                  N   

   pulocationid  dolocationid  payment_type  fare_amount  extra  mta_tax   
0           138           170             1         29.5    1.0      0.5  \
1            11            26             1         10.5    1.0      0.5   
2           163           162             1          9.5    1.0      0.5   
3           170           163             1         13.0    1.0      0.5   
4           163           236             1         10.5    1.0      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount   
0        7.98          6.12                    0.3         47.90  \
1        0.00          0.00                    0.3         12.30   
2        2.00          0.00                    0.3         15.80   
3        4.32          0.00                    0.3         21.62   
4        0.50          0.00                    0.3         15.30   

   congestion_surcharge  
0                   2.5  
1                   0.0  
2                   2.5  
3                   2.5  
4                   2.5

#### Exercise 1.2
##### Find the dimensions (number of rows and number of columns) in the data.

In [27]:
yellow_taxi_trip.loc[(yellow_taxi_trip['trip_distance']>30),['tpep_pickup_datetime','trip_distance']]

tpep_pickup_datetime  trip_distance
1656  2019-10-23T16:04:45.000          37.57
8338  2019-10-23T16:50:53.000          38.11
9965  2019-10-23T17:34:29.000          37.86

#### Exercise 1.3
##### Using the data in the `2019_Yellow_Taxi_Trip_Data.csv` file, calculate summary statistics for the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` columns.

In [28]:
yellow_taxi_trip[['fare_amount','tip_amount','tolls_amount','total_amount']].sum()

fare_amount     151063.13
tip_amount       26344.94
tolls_amount      6234.47
total_amount    225646.59
dtype: float64

#### Exercise 1.4
##### Isolate the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` for the longest trip by distance (`trip_distance`).

In [39]:
## yellow_taxi_trip[yellow_taxi_trip['trip_distance'] == yellow_taxi_trip['trip_distance'].max()][['fare_amount','tip_amount','tolls_amount','total_amount']]

yellow_taxi_trip.nlargest(1,'trip_distance')[['fare_amount','tolls_amount']]

fare_amount  tolls_amount
8338        176.0          6.12

### [Exercise 2.1](./workbook.ipynb#Exercise-2.1)

##### Read in the meteorite data from the `Meteorite_Landings.csv` file, rename the `mass (g)` column to `mass`, and drop all the latitude and longitude columns. Sort the result by mass in descending order.

In [47]:
import pandas as pd

meteorites = pd.read_csv('../data/Meteorite_Landings.csv')

meteorites = meteorites.rename(
    columns={
        'mass (g)': 'mass'
    }
)

## meteorites = meteorites.drop(['latitude','longitude'])
meteorites.head(5)

name   id nametype     recclass      mass  fall   
0    Aachen    1    Valid           L5      21.0  Fell  \
1    Aarhus    2    Valid           H6     720.0  Fell   
2      Abee    6    Valid          EH4  107000.0  Fell   
3  Acapulco   10    Valid  Acapulcoite    1914.0  Fell   
4   Achiras  370    Valid           L6     780.0  Fell   

                     year    reclat    reclong           GeoLocation  
0  01/01/1880 12:00:00 AM  50.77500    6.08333     (50.775, 6.08333)  
1  01/01/1951 12:00:00 AM  56.18333   10.23333  (56.18333, 10.23333)  
2  01/01/1952 12:00:00 AM  54.21667 -113.00000    (54.21667, -113.0)  
3  01/01/1976 12:00:00 AM  16.88333  -99.90000     (16.88333, -99.9)  
4  01/01/1902 12:00:00 AM -33.16667  -64.95000   (-33.16667, -64.95)

### [Exercise 2.2](./workbook.ipynb#Exercise-2.2)

##### Using the meteorite data from the `Meteorite_Landings.csv` file, update the `year` column to only contain the year, convert it to a numeric data type, and create a new column indicating whether the meteorite was observed falling before 1970. Set the index to the `id` column and extract all the rows with IDs between 10,036 and 10,040 (inclusive) with `loc[]`.

###### **Hint 1**: Use `year.str.slice()` to grab a substring.

###### **Hint 2**: Make sure to sort the index before using `loc[]` to select the range.

###### **Bonus**: There's a data entry error in the `year` column. Can you find it? (Don't spend too much time on this.)

In [75]:
import pandas as pd

meteorites = pd.read_csv('../data/Meteorite_Landings.csv')

meteorites = meteorites.rename(
    columns={
        'mass (g)': 'mass'
    }
)

## meteorites = meteorites.drop(['latitude','longitude'])
## meteorites.head(5)

meteorites.year=meteorites.year.str.slice(start=6 ,stop=10).apply(pd.to_numeric)
meteorites['falling_before_1970']=meteorites['year']<1970

meteorites.set_index('id')
meteorites.sort_index
meteorites.loc[10036:10040]



name    id nametype    recclass  mass   fall    year   
10036  Elephant Moraine 90022  8432    Valid         CK5  15.5  Found  1990.0  \
10037  Elephant Moraine 90023  8433    Valid         CK5  31.5  Found  1990.0   
10038  Elephant Moraine 90024  8434    Valid  Eucrite-br  22.8  Found  1990.0   
10039  Elephant Moraine 90025  8435    Valid         CK5  45.8  Found  1990.0   
10040  Elephant Moraine 90026  8436    Valid         CK5  61.5  Found  1990.0   

         reclat    reclong             GeoLocation  falling_before_1970  
10036 -76.28573  156.45721  (-76.28573, 156.45721)                False  
10037 -76.27507  156.41038  (-76.27507, 156.41038)                False  
10038 -76.28843  156.47872  (-76.28843, 156.47872)                False  
10039 -76.28200  156.39926    (-76.282, 156.39926)                False  
10040 -76.29226  156.45353  (-76.29226, 156.45353)                False

#### Exercise 2.3
##### Using the meteorite data from the `Meteorite_Landings.csv` file, create a pivot table that shows both the number of meteorites and the 95th percentile of meteorite mass for those that were found versus observed falling per year from 2005 through 2009 (inclusive). Hint: Be sure to convert the `year` column to a number as we did in the previous exercise.

#### Exercise 2.4
##### Using the meteorite data from the `Meteorite_Landings.csv` file, compare summary statistics of the mass column for the meteorites that were found versus observed falling.

#### Exercise 2.5
##### Using the taxi trip data in the `2019_Yellow_Taxi_Trip_Data.csv` file, resample the data to an hourly frequency based on the dropoff time. Calculate the total `trip_distance`, `fare_amount`, `tolls_amount`, and `tip_amount`, then find the 5 hours with the most tips.

--- 

### Section 3

#### Exercise 3.1
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create box plots for traveler throughput for each year in the data. Hint: Pass `kind='box'` into the `plot()` method to generate box plots.

#### Exercise 3.2
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create a heatmap that shows the 2019 TSA median traveler throughput by day of week and month.

#### Exercise 3.3
##### Annotate the medians in the box plot from *[Exercise 3.1](#Exercise-3.1)*. Hint: The `x` coordinates will be 1, 2, and 3 for 2019, 2020, and 2021, respectively. Alternatively, to avoid hardcoding values, you can use the `Axes.get_xticklabels()` method, in which case you should look at the [documentation](https://matplotlib.org/stable/api/text_api.html) for the `Text` class.